### DISCUSSION TITLE
Modul problem

### DISCUSSION THREAD

**Discussion by paulwwerner** [AllskyTeam/allsky-modules#64] [Q&A]
Unfortunately I have to interrupt again with a problem.

I have my own script for reading out 2 DHT22 sensors:

```
import adafruit_dht
import time

# Definiere Sensortypen und GPIO-Pins
sensor1_type = adafruit_dht.DHT22
sensor1_pin = 13
sensor2_type = adafruit_dht.DHT22
sensor2_pin = 17
sensor1_errors = 0
sensor2_errors = 0

while True:
    try:
        # Werte von den Sensoren auslesen
        time.sleep(10)
        humidity1, temperature1 = adafruit_dht.read_retry(sensor1_type, sensor1_pin)
        if humidity1 is not None and temperature1 is not None:
            sensor1_errors = 0
        else:
            sensor1_errors += 1
        time.sleep(10)
        humidity2, temperature2 = adafruit_dht.read_retry(sensor2_type, sensor2_pin)
        if humidity2 is not None and temperature2 is not None:
            sensor2_errors = 0
        else:
            sensor2_errors += 1

        # Taupunkt berechnen (Beispiel-Formel, durch tatsÃ¤chliche Berechnung ersetzen)
        dew_point1 = round(temperature1 - (100 - humidity1) / 5, 1) if humidity1 is not None and temperature1 is not None else None
        dew_point2 = round(temperature2 - (100 - humidity2) / 5, 1) if humidity2 is not None and temperature2 is not None else None

        # Werte in Datei speichern
        file_path = '/home/pi/allsky/config/overlay/extra/sensor_data.txt'
        with open(file_path, 'w') as file:
            file.write(f'Dome_Temperature={round(temperature1, 1) if temperature1 is not None else "None"}\n')
            file.write(f'Dome_Humidity={round(humidity1, 1) if humidity1 is not None else "None"}\n')
            file.write(f'Dome_Dewpoint={dew_point1 if dew_point1 is not None else "None"}\n')
            file.write(f'Stewa_Temperature={round(temperature2, 1) if temperature2 is not None else "None"}\n')
            file.write(f'Stewa_Humidity={round(humidity2, 1) if humidity2 is not None else "None"}\n')
            file.write(f'Stewa_Dewpoint={dew_point2 if dew_point2 is not None else "None"}\n')
            file.write(f'Sensor1_Errors={sensor1_errors}\n')
            file.write(f'Sensor2_Errors={sensor2_errors}\n')

        time.sleep(30)  # Warte 30 Sekunden, bevor der Code erneut ausgefÃ¼hrt wird
    except Exception as e:
        print(f"Fehler aufgetreten: {e}")
```

It went wonderfully. But since the new Phyton version with Python virtual environment module.

When I start my script I get the following error message:

Traceback (most recent call last):
  File "/home/pi/allsky/klima.py", line 1, in <module>
    import Adafruit_DHT
ModuleNotFoundError: No module named 'Adafruit_DHT'

However, "adafruit_dht" must be installed correctly because the Sky Dew Controler module is running.

Can you help me find my mistake?
Thanks

**Comment by Alex-developer** [comment]
Paul

This is a restriction that has been imposed in the latest version of Bookworm. Python now requires virtual environments 

The good news is its pretty easy to sort out.

Before your code runs you need to activate the python virtual environment. See [this ](https://realpython.com/python-virtual-environments-a-primer/)site for a detailed description on them 

Allsky creates its own Python environment for all of its requirements which you could use bust best practice would be for you to create your own. So assuming you have a user 'pi' try the following

`python3 -m venv /home/pi/myvenv`

This will create the virtual environment. Then install the Adafruit dht22 library in it

`source /home/pi/myvenv/bin/activate`
`pip3 install adafruit-circuitpython-dht`

To use the new Python virtual environment you need to activate it before your script runs so something like


`source /home/pi/myvenv/bin/activate`
`whatever code runs your script`
`deactivate`

There are lots of other ways that you can run the script, such as using shebangs and there are plenty of resource available on the internet that cover this.

Hope that helps

Alex
